src/core/thread/mesh_forwarder_ftd.cpp

```diff
@@ -235,20 +235,38 @@ void MeshForwarder::UpdateIndirectMessages(void)
     }
 }
 
-otError MeshForwarder::EvictIndirectMessage(void)
+otError MeshForwarder::EvictMessage(uint8_t aPriority)
 {
-    otError error = OT_ERROR_NOT_FOUND;
+    otError  error = OT_ERROR_NOT_FOUND;
+    Message *message;
 
-    for (Message *message = mSendQueue.GetHead(); message; message = message->GetNext())
-    {
-        if (!message->IsChildPending())
-        {
-            continue;
-        }
+    VerifyOrExit((message = mSendQueue.GetTail()) != NULL);
 
+    if (message->GetPriority() > aPriority)
+    {
         RemoveMessage(*message);
         ExitNow(error = OT_ERROR_NONE);
     }
+    else
+    {
+        while (1)
+        {
+            for (message = mSendQueue.GetHeadForPriority(aPriority); message; message = message->GetNext())
```

Comment:
When we start from head at a given priority we will iterate till end and possibly going over messages with lower priority that we checked earlier.
How about adding a check in `for` loop to stop when we covered all messages with priority `aPriority`:
```
for (
    message = mSendQueue.GetHeadForPriority(aPriority); 
    message && (message->GetPriority() == aPriority); 
    message = message->GetNext()
)
```

Reply:
Updated.